javascript/tables/src/validationUtils.ts

```diff
@@ -0,0 +1,324 @@
+	return function validateMaximumConstraint(
+		values: unknown,
+		includeInstances: boolean,
+	): ValidationTestResult {
+		const resultIndexes: number[] = []
+		const valuesCasted =
+			dataType === DataType.Number ? (values as number[]) : (values as Date[])
+
+		if (includeInstances) {
+			valuesCasted.forEach((value: number | Date, index: number) => {
+				if (!validate(value)) resultIndexes.push(index)
+			})
+			return {
+				fail: resultIndexes.length !== 0,
+				indexes: resultIndexes,
+				rule: ErrorCode.Maximum,
+			}
+		} else {
+			return {
+				fail: !(dataType === DataType.Number
+					? (values as number[]).every(validate)
+					: (values as Date[]).every(validate)),
+				indexes: resultIndexes,
+				rule: ErrorCode.Maximum,
+			}
+		}
+	}
+}
+
+export function validateMaximum(
+	maximum: number,
+	dataType: DataType,
+): (value: Date | number) => boolean {
+	return function validateMaximum(value: Date | number) {
+		return dataType == DataType.Date
+			? value.valueOf() <= maximum
+			: value <= maximum
+	}
+}
+
+export function validatePatternConstraint(
+	pattern: string,
+): (values: unknown, includeInstances: boolean) => ValidationTestResult {
+	const validate = validatePattern(pattern)
+	return function validatePatternConstraint(
+		values: unknown,
+		includeInstances: boolean,
+	): ValidationTestResult {
+		const resultIndexes: number[] = []
+		const valuesCasted = values as string[]
+
+		if (includeInstances) {
+			valuesCasted.forEach((value: string, index: number) => {
+				if (!validate(value)) resultIndexes.push(index)
+			})
+			return {
+				fail: resultIndexes.length !== 0,
+				indexes: resultIndexes,
+				rule: ErrorCode.Pattern,
+			}
+		} else {
+			return {
+				fail: !valuesCasted.every(validate),
+				indexes: resultIndexes,
+				rule: ErrorCode.Pattern,
+			}
+		}
+	}
+}
+
+export function validatePattern(pattern: string): (value: string) => boolean {
+	return function validatePattern(value: string) {
+		const re = new RegExp(pattern)
+		return re.test(value)
+	}
+}
+
+export function validateEnumConstraint(
+	enumList: string[],
+): (values: unknown, includeInstances: boolean) => ValidationTestResult {
+	const validate = validateEnum(enumList)
+	return function validateEnumConstraint(
+		values: unknown,
+		includeInstances: boolean,
+	): ValidationTestResult {
+		const resultIndexes: number[] = []
+		const valuesCasted = values as string[]
+
+		if (includeInstances) {
+			valuesCasted.forEach((value: string, index: number) => {
+				if (!validate(value)) resultIndexes.push(index)
+			})
+			return {
+				fail: resultIndexes.length !== 0,
+				indexes: resultIndexes,
+				rule: ErrorCode.Enum,
+			}
+		} else {
+			return {
+				fail: !valuesCasted.every(validate),
+				indexes: resultIndexes,
+				rule: ErrorCode.Enum,
+			}
+		}
+	}
+}
+
+export function validateEnum(enumList: string[]): (value: string) => boolean {
+	return function validateEnum(value: string) {
```

Comment:
An optimization you can do here, which shows the cool power of closures: before returning the `validateEnum` function, you can add the `enumList` to a Set. This way the validation check can be `set.has` instead of `enumList.includes`, which will be much faster.

I.e.,

```
  const set = new Set(enumList)
  return function ... {
    return set.has(value)
  }
```

Reply:
The reason closure is cool here is because you only need to create the Set once, and then it is reused by the return function over and over instead of searching the list every time.